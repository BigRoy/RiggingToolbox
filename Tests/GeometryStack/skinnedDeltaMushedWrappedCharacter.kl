
require RiggingToolbox;

operator entry(){

  StartFabricProfiling();

  GeometryStack scrStack();
  scrStack.loadJSONFile("${FABRIC_RIGGINGTOOLBOX_PATH}/Tests/GeometryStack/Resources/tubeCharacter_SkinningAndDeltaMush.json");

  GeometryStack stack();
  stack.loadJSONFile("${FABRIC_RIGGINGTOOLBOX_PATH}/Tests/GeometryStack/Resources/tubeCharacter_Wrap.json");

  WrapModifier wrapModifier = stack.getGeometryOperator(1);
  wrapModifier.setSourceGeomStack(scrStack);

  SkinningModifier skiningModifier = scrStack.getGeometryOperator(1);

  EvalContext context();
  stack.evaluate(context);

  DrawContext drawContext = DrawContext();
  InlineDrawingScope inlineDrawingScope = InlineDrawingScope();
  inlineDrawingScope.draw(drawContext);
  
  // Modify the pose and then reevaluate. 
  {
    Mat44 pose[];
    pose.resize(4);
    pose[0] = Xfo(Vec3(3, 4, 5)).toMat44();
    pose[1] = Xfo(Vec3(10, 20, 5)).toMat44();
    pose[2] = Xfo(Vec3(10, 20, 5)).toMat44();
    pose[3] = Xfo(Vec3(10, 20, 5)).toMat44();
    skiningModifier.setPose(pose);
  }

  stack.evaluate(context);

  inlineDrawingScope.draw(drawContext);


  {
    // Modify the pose again and then reevaluate. 
    Mat44 pose[];
    pose.resize(4);
    pose[0] = Xfo(Vec3(4, 7, 13)).toMat44();
    pose[1] = Xfo(Vec3(14, 16, 17)).toMat44();
    skiningModifier.setPose(pose);
  }
  
  stack.evaluate(context);
  inlineDrawingScope.draw(drawContext);

  
  StopFabricProfiling();

  report( GetEvalPathReport() );

  report("scrStack:" + scrStack.getDesc());
  report("stack:" + stack.getDesc());
  
}

