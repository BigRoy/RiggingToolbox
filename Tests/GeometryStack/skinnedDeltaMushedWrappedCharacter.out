Importing:C:\Users\Phil\Projects\RiggingToolbox/Tests/GeometryStack/Resources\skinnedTube.abc
Importing:C:\Users\Phil\Projects\RiggingToolbox/Tests/GeometryStack/Resources\skinnedTube.abc
AlembicGeometryAttributesReader: Property "EnvelopeWeights" not supported.
AlembicGeometryAttributesReader: Property "MappingIndices" not supported.
AlembicGeometryAttributesReader: Property "EnvelopeDeformerIndices" not supported.
AlembicGeometryAttributesReader: Property "EnvelopeDeformers" not supported.
Warning: Pose count does not match the reference pose count. referencePose:4 != pose:0. Skinning disabled
function Ref<GeometryStack>#notify!(Notifier, String, String):WrapModifier.changed
function GeometrySet Ref<GeometryStack>.evaluate!(EvalContext)
--function Ref<GeometryCache>#restore?(io GeometrySet)
--function Ref<AlembicGeometryGenerator>#evaluate!(EvalContext, io GeometrySet)
--function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):["positions"]
--function Ref<WrapModifier>#evaluate!(EvalContext, io GeometrySet)
----function GeometrySet Ref<GeometryStack>.evaluate!(EvalContext)
------function Ref<GeometryCache>#restore?(io GeometrySet)
------function Ref<AlembicGeometryGenerator>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):["positions"]
------function Ref<SkinningModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):["positions"]
------function Ref<DeltaMushModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
------function Ref<ComputeNormalsModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
------function Ref<ComputeTangentsModifier>#evaluate!(EvalContext, io GeometrySet)
--function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
--function Ref<ComputeNormalsModifier>#evaluate!(EvalContext, io GeometrySet)
function Ref<GeometryStack>#notify!(Notifier, String, String):SkinningModifier.changed
--function Ref<WrapModifier>#notify!(Notifier, String, String):GeometryStack.changed
----function Ref<GeometryStack>#notify!(Notifier, String, String):WrapModifier.changed
function GeometrySet Ref<GeometryStack>.evaluate!(EvalContext)
--function Ref<GeometryAttributeCache>#restore?(io GeometrySet):["positions"]
--function Ref<WrapModifier>#evaluate!(EvalContext, io GeometrySet)
----function GeometrySet Ref<GeometryStack>.evaluate!(EvalContext)
------function Ref<GeometryAttributeCache>#restore?(io GeometrySet):["positions"]
------function Ref<SkinningModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):["positions"]
------function Ref<DeltaMushModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
------function Ref<ComputeNormalsModifier>#evaluate!(EvalContext, io GeometrySet)
------function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
------function Ref<ComputeTangentsModifier>#evaluate!(EvalContext, io GeometrySet)
--function Ref<GeometryAttributeCache>#update!(GeometrySet, GeometryOperator):[]
--function Ref<ComputeNormalsModifier>#evaluate!(EvalContext, io GeometrySet)

scrStack:GeometryStack { 
  geomOperators:[ 
  {
    type: AlembicGeometryGenerator,
    filePath: skinnedTube.abc
    expandedPath: C:\Users\Phil\Projects\RiggingToolbox/Tests/GeometryStack/Resources\skinnedTube.abc
  }
  { 
    type: SkinningModifier
  }
  { 
    type: DeltaMushModifier
  }
  { 
    type: ComputeNormalsModifier
  }
  { 
    type: ComputeTangentsModifier
  }
  ],
  numGeometries:1
}
stack:GeometryStack { 
  geomOperators:[ 
  {
    type: AlembicGeometryGenerator,
    filePath: skinnedTube.abc
    expandedPath: C:\Users\Phil\Projects\RiggingToolbox/Tests/GeometryStack/Resources\skinnedTube.abc
  }
  { 
    type: WrapModifier
  }
  { 
    type: ComputeNormalsModifier
  }
  ],
  numGeometries:1
}
